ckanext/scheming/helpers.py

```diff
@@ -343,3 +347,46 @@ def scheming_display_json_value(value, indent=2):
         return json.dumps(value, indent=indent, sort_keys=True)
     except (TypeError, ValueError):
         return value
+
+
+def scheming_render_from_string(source, **kwargs):
+    # Temporary solution for rendering defaults and including the CKAN
+    # helpers. The core CKAN lib does not include a string rendering
+    # utility that works across 2.6-2.8.
+    from ckantoolkit import h
+
+    env = Environment(autoescape=True)
+    template = env.from_string(
+        source,
+        globals={
+            'h': h
+        }
+    )
+
+    return template.render(**kwargs)
+
+
+def scheming_massage_subfield(field, subfield, index, data):
+    # Subfield & data are re-used, we must not modify the original!
+    sf = subfield.copy()
+    data = data.copy()
+
+    new_field_name = '{field_name}-{index}-{subfield_name}'.format(
```

Comment:
I learned that our controllers already have a convention for nested parameters: 
https://github.com/ckan/ckan/blob/2.7/ckan/controllers/package.py#L564-L565

This converts parameters in the form `field_name__index__subfield_name` into lists of nested dicts automatically in the controller. We should use this instead of inventing another nested list-of-dicts implementation for scheming.

Reply:
This could be too large a change to ask for as part of getting this PR merged. Let's just revisit later.

Reply:
It's commented elsewhere, but it's done this way because that's already how composite and repeating do it. Without this, plugins that were built on top of them cannot find the fields.